File: dockerfiles/attestation-sidecar-py/app/internal/storage/memory.py
# app/internal/storage/memory.py

import threading
import json
import os
from typing import Any, Dict, Optional
from app.internal.logger import info, warn, error, debug, green_text

class MemoryStorageBackend:
    def __init__(self):
        self._data: Dict[str, Any] = {}
        self._lock = threading.Lock()
        self._storage_file = os.getenv("MEMORY_STORAGE_FILE", "")
        
        # Charger les données depuis le fichier si spécifié
        # if self._storage_file and os.path.exists(self._storage_file):
        #     self.load_from_file(self._storage_file)

    def unpack_set(self, **kwargs: Any) -> None:
        with self._lock:
            self._data.update(kwargs)

    def set(self, key: str, value: Any) -> None:
        with self._lock:
            self._data[key] = value

    def get(self, key: str) -> Optional[Any]:
        with self._lock:
            return self._data.get(key)

    def delete(self, key: str) -> None:
        with self._lock:
            if key in self._data:
                del self._data[key]

    def exists(self, key: str) -> bool:
        with self._lock:
            return key in self._data
            
    def set_status(self, key: str, value: str) -> None:
        self.set(key, value)
    
    def store_agent_info(self, key: str, info: Dict[str, Any]) -> None:
        self.set(key, info)
    
    def push_final_digest(self, key: str, digest_data: Dict[str, Any]) -> None:
        if key not in self._data:
            with self._lock:
                self._data[key] = []
        with self._lock:
            self._data[key].append(digest_data)
    
    def incr_digest_count(self, key: str) -> int:
        with self._lock:
            if key not in self._data:
                self._data[key] = 0
            self._data[key] += 1
            self._save_to_file_if_configured()
            return self._data[key]
    
    def export_to_file(self, file_path: str) -> bool:
        """Exporte toutes les données du stockage vers un fichier JSON"""
        try:
            with self._lock:
                with open(file_path, 'w') as f:
                    json.dump(self._data, f, indent=2)
            debug(f"Storage data exported to {file_path}")
            return True
        except Exception as e:
            error(f"Failed to export storage data to {file_path}: {str(e)}")
            return False
    
    def load_from_file(self, file_path: str) -> bool:
        """Charge les données depuis un fichier JSON vers le stockage"""
        try:
            with open(file_path, 'r') as f:
                data = json.load(f)
            with self._lock:
                self._data = data
            info(f"Storage data loaded from {file_path}")
            return True
        except Exception as e:
            error(f"Failed to load storage data from {file_path}: {str(e)}")
            return False
    
    def _save_to_file_if_configured(self):
        """Sauvegarde les données dans le fichier configuré si spécifié
        Optimisé pour ne pas sauvegarder à chaque incrément (seulement tous les 10)
        """
        if not self._storage_file:
            return
            
        # Récupérer ou initialiser le compteur de modifications
        with self._lock:
            mod_count = self._data.get('_mod_count', 0) + 1
            self._data['_mod_count'] = mod_count
            
            # Sauvegarder seulement tous les 10 incréments pour éviter trop d'I/O
            if mod_count % 10 == 0:
                try:
                    debug(f"Saving memory storage to {self._storage_file} (modification count: {mod_count})")
                    self.export_to_file(self._storage_file)
                except Exception as e:
                    error(f"Failed to auto-save memory storage: {str(e)}")
                    # Continuer l'exécution même en cas d'erreur

File: dockerfiles/attestation-sidecar/app/internal/storage/memory.go
package storage

import (
	"encoding/json"
	"os"
	"sync"

	"app/internal/logger"
)

// MemoryStorageBackend implements in-memory storage
type MemoryStorageBackend struct {
	data        map[string]interface{}
	mu          sync.RWMutex
	storageFile string
	modCount    int
}

// NewMemoryStorageBackend creates a new memory storage backend
func NewMemoryStorageBackend() *MemoryStorageBackend {
	return &MemoryStorageBackend{
		data:        make(map[string]interface{}),
		storageFile: os.Getenv("MEMORY_STORAGE_FILE"),
	}
}

func (m *MemoryStorageBackend) Set(key string, value interface{}) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.data[key] = value
	return nil
}

func (m *MemoryStorageBackend) Get(key string) (interface{}, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.data[key], nil
}

func (m *MemoryStorageBackend) Delete(key string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.data, key)
	return nil
}

func (m *MemoryStorageBackend) Exists(key string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	_, exists := m.data[key]
	return exists
}

func (m *MemoryStorageBackend) SetStatus(key, value string) error {
	return m.Set(key, value)
}

func (m *MemoryStorageBackend) StoreAgentInfo(key string, info map[string]interface{}) error {
	return m.Set(key, info)
}

func (m *MemoryStorageBackend) PushFinalDigest(key string, digestData map[string]interface{}) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	var list []map[string]interface{}
	if existing, ok := m.data[key]; ok {
		if existingList, ok := existing.([]map[string]interface{}); ok {
			list = existingList
		}
	}

	list = append(list, digestData)
	m.data[key] = list
	return nil
}

func (m *MemoryStorageBackend) IncrDigestCount(key string) (int, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	count := 0
	if existing, ok := m.data[key]; ok {
		if existingCount, ok := existing.(int); ok {
			count = existingCount
		}
	}

	count++
	m.data[key] = count

	// Auto-save every 10 increments
	m.modCount++
	if m.modCount%10 == 0 && m.storageFile != "" {
		m.saveToFileIfConfigured()
	}

	return count, nil
}

func (m *MemoryStorageBackend) ExportToFile(filePath string) error {
	m.mu.RLock()
	defer m.mu.RUnlock()

	data, err := json.MarshalIndent(m.data, "", "  ")
	if err != nil {
		logger.Error("Failed to marshal storage data: %v", err)
		return err
	}

	if err := os.WriteFile(filePath, data, 0644); err != nil {
		logger.Error("Failed to export storage data to %s: %v", filePath, err)
		return err
	}

	logger.Debug("Storage data exported to %s", filePath)
	return nil
}

func (m *MemoryStorageBackend) LoadFromFile(filePath string) error {
	data, err := os.ReadFile(filePath)
	if err != nil {
		logger.Error("Failed to load storage data from %s: %v", filePath, err)
		return err
	}

	m.mu.Lock()
	defer m.mu.Unlock()

	if err := json.Unmarshal(data, &m.data); err != nil {
		logger.Error("Failed to unmarshal storage data: %v", err)
		return err
	}

	logger.Info("Storage data loaded from %s", filePath)
	return nil
}

func (m *MemoryStorageBackend) saveToFileIfConfigured() {
	if m.storageFile == "" {
		return
	}

	// Note: caller should already hold the lock
	data, err := json.MarshalIndent(m.data, "", "  ")
	if err != nil {
		logger.Error("Failed to auto-save memory storage: %v", err)
		return
	}

	if err := os.WriteFile(m.storageFile, data, 0644); err != nil {
		logger.Error("Failed to auto-save memory storage: %v", err)
	}
}
